### PR TITLE
User model created using Devise gem

### PR DESCRIPTION
**Pull request by hav3loc**
We created model User using Devise gem.

### CODE REVIEW

**Comment by hav3loc** [comment]
+1

File: Gemfile [hunk 1]
```diff
@@ -10,12 +10,15 @@ gem 'turbolinks', '~> 5'
 gem 'jbuilder', '~> 2.5'
 
 gem 'slim'
-gem 'bootstrap-sass', '~> 3.3.6'
+gem 'rails-assets-hocus-pocus', source: 'https://rails-assets.org'
```
Comment (gotar) [inline Gemfile]: why you have this code (line in you PR)? it's not related

File: Gemfile [hunk 2]
```diff
@@ -10,12 +10,15 @@ gem 'turbolinks', '~> 5'
...
 
 gem 'devise'
 gem 'simple_form'
 gem 'pundit'
 
+group :test do
+  gem 'shoulda-matchers', '~>3.1'
+end
 group :development, :test do
```
Comment (gotar) [inline Gemfile]: missing empty line (not a big fun of shoulda but if you like you can use it ;])

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class User < ApplicationRecord
+  has_many :events
```
Comment (gotar) [inline app/models/user.rb]: there is no events db, why you add relation between them

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
+
+  validates :name, presence: true
+  validates :surname, presence: true
+  validates :city, presence: true
+  validates :name, format: /[A-Z][a-z]*/
```
Comment (gotar) [inline app/models/user.rb]: combine those two lines related to :name

File: app/views/devise/unlocks/new.html.slim [hunk 5]
```diff
@@ -0,0 +1,10 @@
+h2
+  | Resend unlock instructions
+= simple_form_for(resource, as: resource_name, url: unlock_path(resource_name), html: { method: :post }) do |f|
+  = f.error_notification
+  = f.full_error :unlock_token
+  .form-inputs
+    = f.input :email, required: true, autofocus: true
+  .form-actions
+    = f.button :submit, "Resend unlock instructions"
+= render "devise/shared/links"
```
Comment (gotar) [inline app/views/devise/unlocks/new.html.slim]: if they are generic devise view you don't need them. Just drop those files

File: config/initializers/devise.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+Devise.setup do |config|
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
Comment (gotar) [inline config/initializers/devise.rb]: ??

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  it "has many events" do
+    relation=User.reflect_on_association(:events)
+    expect(relation.macro).to equal :has_many
+  end
+  it { should validate_presence_of(:name) }
```
Comment (gotar) [inline spec/models/user_spec.rb]: missing empty line

File: app/models/user.rb (hunk 3 above)
Comment (KamilFilip) [inline app/models/user.rb]: It was to be next inline to add. I thought its my resposnibility to include this now. Same goes for testing of this association, I wrote it as pending but changed to reflect this.

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+  it { should validate_presence_of(:surname) }
+  it { should validate_presence_of(:city) }
```
Comment (gotar) [inline spec/models/user_spec.rb]: if you decide to use should you should test all validation not only chosen one

File: spec/rails_helper.rb [hunk 9]
```diff
@@ -6,6 +6,12 @@
 
 ActiveRecord::Migration.maintain_test_schema!
 
+Shoulda::Matchers.configure do |config|
+  config.integrate do |with|
+    with.test_framework :rspec
+    with.library :rails
+  end
+end
 RSpec.configure do |config|
```
Comment (gotar) [inline spec/rails_helper.rb]: missing empty line

File: Gemfile (hunk 1 above)
Comment (KamilFilip) [inline Gemfile]: I guess it was changed due to merge to my branch

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -1,11 +1,8 @@
 require 'rails_helper'
 
 RSpec.describe User, type: :model do
-  it "has many events" do
-    relation=User.reflect_on_association(:events)
-    expect(relation.macro).to equal :has_many
-  end
+
```
Comment (Snieezy) [inline spec/models/user_spec.rb]: drop empty line

Comment (maciek1995) [inline spec/models/user_spec.rb]: shouldn't you drop this line?

**Comment by maciek1995** [comment]
+1

File: app/models/user.rb [hunk 11]
```diff
@@ -1,12 +1,8 @@
...
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
 
-  validates :name, presence: true
-  validates :surname, presence: true
-  validates :city, presence: true
-  validates :name, format: /[A-Z][a-z]*/
-  validates :surname, format: /[A-Z][a-z]*/
+  validates :name, :surname, :city, presence: true
+  validates :surname, :surname format: /[A-Z][a-z]*/
```
Comment (hav3loc) [inline app/models/user.rb]: :surname, :surname  ?

Comment (KamilFilip) [inline app/models/user.rb]: Poprawione, sorki niedopatrzenie

**Comment by smt116** [comment]
Why have you used `git merge` to update your branch instead of `git rebase`? Why there are 14 commits and most of them with really poor description?

File: Gemfile (hunk 1 above)
Comment (gotar) [inline Gemfile]: so fix it ;]

File: Gemfile [hunk 12]
```diff
@@ -10,12 +10,16 @@ gem 'turbolinks', '~> 5'
...
 gem 'slim'
-gem 'bootstrap-sass', '~> 3.3.6'
+gem 'rails-assets-hocus-pocus', source: 'https://rails-assets.org'
 
 gem 'devise'
 gem 'simple_form'
 gem 'pundit'
 
+group :test do
+  gem 'shoulda-matchers', '~>3.1'
```
Comment (gotar) [inline Gemfile]: not a big fan of shoulda and testing validations from model

File: app/models/user.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
+
+  validates :name, :surname, :city, presence: true
+  validates :name, :surname, format: /[A-Z][a-z]*/
```
Comment (gotar) [inline app/models/user.rb]: from regexp you allow for empty name and surename, I think you want + in place of *. 

ps If user name 2 names or number in name or other strange stuff?, just present should be enough

File: spec/models/user_spec.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  it { should validate_presence_of(:name) }
+  it { should validate_presence_of(:surname) }
+  it { should validate_presence_of(:city) }
```
Comment (gotar) [inline spec/models/user_spec.rb]: its okish but not a good fan of it

Comment (Fredar) [inline spec/models/user_spec.rb]: I would update the syntax to `it { is_expected.to validate_presence_of(:name) }`

File: Gemfile (hunk 12 above)
Comment (gotar) [inline Gemfile]: but leave it if you like to learn how it works

**Comment by gotar** [comment]
fix commits history, squash

**Comment by gotar** [comment]
squash and you can merge

**Pull request merged**